core/src/main/scala/kafka/raft/KafkaMetadataLog.scala

```diff
@@ -315,7 +324,8 @@ object KafkaMetadataLog {
       }
       .forEach { path =>
         path.ifPresent { snapshotPath =>
-          if (snapshotPath.partial) {
+          if (snapshotPath.partial || snapshotPath.snapshotId.offset < log.logStartOffset) {
```

Comment:
Right. I think we need to do this because the `RaftClient` can fail in `deleteBeforeSnapshot` after updating the log start offset but before deleting the snapshots older than the log start offset. Does this match your thinking?

Reply:
Here I am thinking of uncleaned broker shutdown, the case you describe is one of them.